### DISCUSSION TITLE
Funktionsweise der fwmod_custom

### DISCUSSION THREAD

**Discussion by JosKabe** [Freetz-NG/freetz-ng#859] [Allgemein] [answered]
Hallo!

Früher unter freetz (non ng) hatte ich mit der _fwmod_custom_ zusätzlich eigene Modifikationen hinzugefügt, das funktionierte ohne Probleme. Nun habe ich mich an freetz-ng herangetraut und musste dabei feststellen, das meine zusätzlichen Änderungen aus der _fwmod_custom_ (aus dem Abschnitt `all()`) offenbar nicht angewendet werden.

Die Änderung vom 10.09.2023:
https://github.com/Freetz-NG/freetz-ng/commit/11af734cfc67d5ca00bbc092d94a7f10e4fc3a30
hatte ich mitbekommen aber da ich bisher schon ${FILESYSTEM_MOD_DIR} nutzte dürfte das nicht die Ursache sein. Zum Test habe ich es auch mit "build/modified/filesystem/" und "${build_dir}/modified/filesystem/" probiert was aber beides nicht half.

Hier ein diff zwischen der originalen _fwmod_custom_ und meiner ergänzten:
```
--- fwmod_custom.original
+++ fwmod_custom.my
@@ -8,7 +8,16 @@
 #  build/modified/kernel/	- unpacked kernel & hidden kernel squashfs
 
 all() {
-	dummy=0
+#	dummy=0
+#
+##	SIGNING-KEYS
+	cp $HOME/freetz/mymods/filesystem/etc/my_key1 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key1
+	cp $HOME/freetz/mymods/filesystem/etc/my_key2 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key2
+#	cp $HOME/freetz/mymods/filesystem/etc/my_key3 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key3
+	cp $HOME/freetz/mymods/filesystem/etc/my_key4 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key4
+	cp $HOME/freetz/mymods/filesystem/etc/my_key7 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key7
+	cp $HOME/freetz/mymods/filesystem/etc/avm_firmware_public_key9 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key9
+##
 #	local build_dir="$(dirname "$0")/build"
 #	# Use AVM's BusyBox, many Freetz functions will not work!
 #	cp -p ${build_dir}/original/filesystem/bin/busybox ${build_dir}/modified/filesystem/bin/busybox
```

Aber vielleicht habe ich was anderes übersehen was sich seit dem alten freetz geändert hat und die Verwendung der fwmod_custom betrifft.

Build-System ist ein Debian 12 als auch ein Arch-Linux.  Problem besteht bei beiden Systemen und es ist auch egal für welches FRITZ!Box-Modell ich die Firmware bauen möchte.

Falls weitere Informationen (Log, config usw.) benötigt werden werde ich diese gerne liefern.

**Comment by JosKabe** [comment]
Ich habe mir das noch einmal in Ruhe angesehen und den Bereich in _tools/freetz_functions_ aufgerufen der die fwmod_custom aufrufen soll und auch nach älteren Images von mir gesucht.

Ich hatte nach freetz schon einmal mit freetz-ng gearbeitet, am 18.11.2021 (8d8d6f5d0, 2021-11-17), die Images hatte ich bisher aber nicht verwendet. Die damals von mir in der _fwmod_custom_ hinzugefügten Modifikationen sind in diesen alten freetz-ng Images aus 2021 allerdings enthalten, also muss es da noch funktioniert haben.
Seitdem gab es in der _tools/freetz_functions_ im Bereich des _fwmod_custom_ Aufruf imho aber keine relevanten Änderungen. Die einzige auffällige ist die vom 09.09.2023:
https://github.com/Freetz-NG/freetz-ng/commit/18fe85caa26194899ed1a5c7baaadc95aa83f1df

Da ich aber auch schon am 04.09.2023 mit _22364_ 945a29e28 Images erstellte, also noch vor der Änderung und darin meine Modifikationen aus der _fwmod_custom_ in den gebauten Images auch nicht enthalten sind, muss es nicht daran liegen. Wobei ich da die Funktion meiner Aufrufe in der _fwmod_custom_ nicht extra getestet hatte. Dummerweise habe ich die am 04.09. verwendete _fwmod_custom_ nicht mehr, da kann es also auch ein Fehler darin gewesen sein. Ist mir leider erst vor ein paar Tagen aufgefallen.

Da ich nicht weiß ob das vielleicht ein Fehler in freetz-ng ist oder ich einen Fehler in der fwmod_custom gemacht habe, habe ich es noch einmal hardcoded probiert und die Pfade überprüft, in der _fwmod_custom_ und auch ein direkter Aufruf der Befehle aus dem Basisverzeichnis des Checkouts, also z.B.:
`cp -p /home/user/freetz/mymods/filesystem/etc/my_key1 build/modified/filesystem/etc/avm_firmware_public_key1`

Anstatt:
`cp $HOME/freetz/mymods/filesystem/etc/my_key1 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key1`
oder
`cp -p $HOME/freetz/mymods/filesystem/etc/my_key1 ${FILESYSTEM_MOD_DIR}/etc/avm_firmware_public_key1`

Der Aufruf des Befehl direkt im Terminal vom Basisverzeichnis des freetz-ng Checkouts aus funktioniert und die gewünschte Änderung ist dann auch unter _build/modified/filesystem/etc_ zu finden aber bei einem normalen _make_ leider nach wie vor nicht, so als würde die _fwmod_custom_ dabei nicht aufgerufen. Ich konnte bei der Ausgabe auch keine Fehlermeldung finden.


_edit_
Ich habe zum Test eine alte freetz-ng Version ausgecheckt, Tag [ng22040](https://github.com/Freetz-NG/freetz-ng/releases/tag/ng22040), damit funktioniert es!

Bei freetz-ng [ng22040](https://github.com/Freetz-NG/freetz-ng/releases/tag/ng22040) wird unter **STEP 2: MODIFY** zwischen _"processing kernel"_ und _"processing mod base options"_ auch _"invoking custom scripts"_ ausgegeben, das fehlt bei freetz-ng 22609 e0349fc1e.


_edit 2_
Nun habe ich noch freetz-ng [ng22100](https://github.com/Freetz-NG/freetz-ng/releases/tag/ng22100) getestet, damit geht es schon **nicht** mehr. Es fehlt auch _"invoking custom scripts"_ in der Ausgabe unter STEP 2.


_edit 3_
freetz-ng [ng22070](https://github.com/Freetz-NG/freetz-ng/releases/tag/ng22040) kann ich nicht testen, da mir dazu die `mips_gcc-5.5.0-freetz-r19900-shared-glibc.tar.lzma` im dl-Ordner fehlt und ich im Web auch keine andere Quelle dafür finden konnte. Bleibt festzuhalten, der Fehler ist wohl irgendwo zwischen Tag [ng22040](https://github.com/Freetz-NG/freetz-ng/releases/tag/ng22040) und [ng22100](https://github.com/Freetz-NG/freetz-ng/releases/tag/ng22100) dazugekommen.


_edit 4_
Die einzigen Änderungen an der _freetz_functions_ (https://github.com/Freetz-NG/freetz-ng/commits/master/tools/freetz_functions) zwischen ng22040 und ng22100 sind die drei Commits:
https://github.com/Freetz-NG/freetz-ng/commit/bea7cc7a20a8d59eded57ab94166a4bd29bb5423
https://github.com/Freetz-NG/freetz-ng/commit/71b269f5cbcc3f090107636555fc1c4496ed39d8 und
https://github.com/Freetz-NG/freetz-ng/commit/5fd83a38b62605a5f4bc27a8a937d1d722265c57
Aber bei allen dreien fehlt mir gerade die Idee das damit in Zusammenhang zu bringen.

**Comment by fda77** [comment]
Das git deines System wurde geupdated und mag kein Fakreroot/Pseudo mehr
```<
+ local dirty
++ git ls-files -m ./fwmod_custom
++ wc -l
Schwerwiegend: dubiose Besitzverhältnisse im Repository bei '/home/freetz-ng/freetz-ng' entdeckt
Um eine Ausnahme für dieses Verzeichnis hinzuzufügen, rufen Sie auf:

        git config --global --add safe.directory /home/freetz-ng/freetz-ng
+ '[' 0 '!=' 0 ']'
```
Mal wieder ein Fall für https://git-man-page-generator.lokaltog.net/

> da mir dazu die mips_gcc-5.5.0-freetz-r19900-shared-glibc.tar.lzma

Es gibt keinen Mirror mit Platz, einfach auf toolchain selbst bauen umstellen

**Reply by fda77** [reply]
Hier kommt der Schwachsinn her: https://github.blog/2022-04-18-highlights-from-git-2-36/#stricter-repository-ownership-checks

**Comment by fda77** [comment, marked as answer]
* 64414202f invoke_fwmod_custom: remove alle checks
